Framework - Resource Identity Documentation (#361)
hashicorp/web-unified-docs: content/terraform-plugin-framework/v1.15.x/docs/plugin/framework/resources/identity.mdx

---
page_title: Identity
description: >-
Learn how to implement resource identity in the Terraform plugin framework.
---

# Resource identity

A resource identity is a data object, determined by the provider, that is stored alongside the resource state to uniquely identify a remote object. A resource identity should have the following properties:
- The resource identity must correspond to at most one remote object per provider, across all instances of that provider.
- Given a resource identity (during `import`), the provider must be able to determine whether the corresponding remote object exists, and if so, return the resource state.
- Unless the resource identity schema is being upgraded, the identity data must not change during the lifecycle of the remote object.

> - That is, from the creation of the remote object in the remote system until its destruction.

Reviewer: I added this bullet point above because I felt like they were covering the same thing so you can delete this one!
```suggestion
```

## Resource Identity Schema

> To define the identity object for a managed resource, a resource identity [schema](/terraform/plugin/framework/handling-data/schemas) is provided that consists of a map of attribute names and associated behaviors. This is similar to the existing resource state schema, with a few key differences:

Reviewer: Removing this sentence to below to start the list
```suggestion
To define the identity object for a managed resource, a resource identity [schema](/terraform/plugin/framework/handling-data/schemas) is provided that consists of a map of attribute names and associated behaviors.
```

> 1. The identity schema only supports primitive types and list types such as:

Reviewer: ```suggestion
The resource identity schema is similar to the resource state schema, but with the following differences:
1. The identity schema only supports primitive types and list types such as:
```

Is this list exhaustive? Are there things missing from this list?

Author: Yeah it's just these two items at the moment 👍🏻 (in-terms of how the identity schema differs from other schemas)

    - `BoolAttribute`
    - `Float32Attribute`
    - `Float64Attribute`
    - `Int32Attribute`
    - `Int64Attribute`
    - `NumberAttribute`
    - `StringAttribute`
    - `ListAttribute`
        - `ElementType: types.BoolType`
        - `ElementType: types.Float32Type`
        - `ElementType: types.Float64Type`
        - `ElementType: types.Int32Type`
        - `ElementType: types.Int64Type`
        - `ElementType: types.NumberType`
        - `ElementType: types.StringType`

> <Note>

Reviewer: I'd strongly encourage removing the note, [according to our style guide](https://docs.google.com/document/d/1gE5FWf_eEqoeUw4kWUht1YVucbVPXFMxnYNQ6ndlPpc/edit?tab=t.ssj87oswdn7h#bookmark=id.2zqlmzl3r7k1) we try to avoid adding callouts unless they are absolutely necessary.
```suggestion
```

(If you accept this please remove the rest of the lines below up to line 40

Author: Good catch, removed in 4177307aac46433d230de9eed10a6ef3488366a4 ⚡

    The ordering in lists is expected to be stable and semantically relevant.

</Note>

> 2. Unlike the resource state schema, the resource identity schema will not support behaviors such as Required and Computed. Two behaviors will be allowed for each attribute: `RequiredForImport` and `OptionalForImport`, which, as named, only affect behavior when importing a resource by identity.

Reviewer: Always use markdown numbers in lists (`1.`) to avoid having to manually number things later. 
```suggestion
1. Unlike the resource state schema, the resource identity schema does not support behaviors such as `Required` and `Computed`. Two behaviors are allowed for each resource state schema: `RequiredForImport` and `OptionalForImport`, which affect behavior when importing a resource by identity.
```

Example of an identity schema for an aws_s3_bucket:
```go
var _ resource.ResourceWithIdentity = &S3BucketResource{} // new interface

func (r *S3BucketResource) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
	resp.TypeName = "aws_s3_bucket"
}

func (r *S3BucketResource) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
	resp.Schema = schema.Schema{
		// .. existing resource schema for aws_s3_bucket
	}
}

func (r *S3BucketResource) IdentitySchema(_ context.Context, _ resource.IdentitySchemaRequest, resp *resource.IdentitySchemaResponse) {
	resp.IdentitySchema = identityschema.Schema{
		Attributes: map[string]identityschema.Attribute{
			"bucket": identityschema.StringAttribute{
				RequiredForImport: true,
			},
			"account_id": identityschema.StringAttribute{
				OptionalForImport: true,
			},
			"partition": identityschema.StringAttribute{
				OptionalForImport: true,
			},
			"region": identityschema.StringAttribute{
				OptionalForImport: true,

			},
		},
	}
}
```

## Identity Schema Usage

> 

Reviewer: Can you please add a sentence here to avoid stacking headings?

Author: Will remove the stack headings, need to refactor this section a bit first 👍🏻

A resource identity should have the following properties:
- The resource identity must correspond to at most one remote object per provider, across all instances of that provider.
- Given a resource identity (during `import`), the provider must be able to determine whether the corresponding remote object exists, and if so, return the resource state.
- The identity data for a remote object must not change during its lifecycle from creation to deletion, or until the provider [upgrades](/terraform/plugin/framework/resources/identity-upgrade) the resource identity schema.

## Schema

To define the identity object for a managed resource, a resource identity [schema](/terraform/plugin/framework/handling-data/schemas) is provided that consists of a map of attribute names and associated behaviors.

The resource identity schema is similar to the resource state schema, but with the following differences:
- The identity schema only supports primitive types and list types, which are represented by the following attributes:

> | Attribute                                                                          | ElementType                                                                    |

Reviewer 2: Nice, I like the table

  resp.Schema = schema.Schema{
    // .. resource schema for examplecloud_thing
  }
}

// Struct model for identity data handling
type ThingResourceIdentityModel struct {
  ID        types.String `tfsdk:"id"`
  Region    types.String `tfsdk:"region"`
}

func (r ThingResource) IdentitySchema(_ context.Context, _ resource.IdentitySchemaRequest, resp *resource.IdentitySchemaResponse) {
  resp.IdentitySchema = identityschema.Schema{
    Attributes: map[string]identityschema.Attribute{
      "id": identityschema.StringAttribute{
        RequiredForImport: true, // must be set during import by the practitioner
      },
      "region": identityschema.StringAttribute{
        OptionalForImport: true, // can be defaulted by the provider configuration
      },
    },
  }
}
```

## Handling Identity Data

Identity data, similar to resource state data, can be set or retrieved during the resource [`Create`](/terraform/plugin/framework/resources/create),
[`Read`](/terraform/plugin/framework/resources/read), [`Update`](/terraform/plugin/framework/resources/update), [`Delete`](/terraform/plugin/framework/resources/delete)

> and [`ImportState`](/terraform/plugin/framework/resources/import) methods. Unlike resource state data, identity data is expected to be immutable after it is set during

Reviewer 2: nit: We could mention or link to mutable identities here, but not necessary. I could see how mentioning that might make people think it is more normal than it is for identity to change.

Author: Yeah I think the latter portion of your thinking is where I'm at. I'd prefer to keep that information (while available), not the norm 😆 

I'm gonna leave it here for now, if we feel like we need to be more explicit/forward in the future we can always come back and adjust 👍🏻

---
page_title: Identity
description: >-
Learn how to implement resource identity in the Terraform plugin framework.
---

# Resource identity

> A resource identity is a data object that is stored alongside the resource instance state to refer to at most one remote object, across all instances of that provider.

Author: This feels like the best place to get this explanation out of the way 😆. Eventually we can add more words about the bad stuff that happens when identity changes, but this feels good for now.

```suggestion
A resource identity is a data object, determined by the provider, that is stored alongside the resource state to uniquely identify a remote object. A resource identity should have the following properties:
- The resource identity must correspond to at most one remote object per provider, across all instances of that provider.
- Given a resource identity (during `import`), the provider must be able to determine whether the corresponding remote object exists, and if so, return the resource state.
- Unless the resource identity schema is being upgraded, the identity data must not change during the lifecycle of the remote object.
  - That is, from the creation of the remote object in the remote system until its destruction.
```

## Resource Identity Schema

> Practitioners can create an resource identity schema, like the existing managed resource schema, that consists of a map of attribute names and associated behaviours. With a few key differences:

Author: ```suggestion
To define the identity object for a managed resource, a resource identity [schema](/terraform/plugin/framework/handling-data/schemas) is provided that consists of a map of attribute names and associated behaviors. This is similar to the existing resource state schema, with a few key differences:
```

> 1. The identity schema will only support primitive types and list types such as:

Author: ```suggestion
1. The identity schema only supports primitive types and list types such as:
```

    - `BoolAttribute`
    - `Float32Attribute`
    - `Float64Attribute`
    - `Int32Attribute`
    - `Int64Attribute`
    - `NumberAttribute`
    - `StringAttribute`
    - `ListAttribute`
        - `ElementType: types.BoolType`
        - `ElementType: types.Float32Type`
        - `ElementType: types.Float64Type`
        - `ElementType: types.Int32Type`
        - `ElementType: types.Int64Type`
        - `ElementType: types.NumberType`
        - `ElementType: types.StringType`

> NOTE: The ordering in lists is expected to be stable and semantically relevant.

Author: ```suggestion
<Note>

The ordering in lists is expected to be stable and semantically relevant.

</Note>
```

> 2. Unlike the resource config schema, the resource identity schema will not support behaviours such as Required and Computed. Two new, but similar, behaviours will be allowed for each attribute: `RequiredForImport` and `OptionalForImport`.

Author: Nit: don't want to make any false equivalence between the existing behaviors and these new ones 😄 

```suggestion
2. Unlike the resource state schema, the resource identity schema will not support behaviors such as Required and Computed. Two behaviors will be allowed for each attribute: `RequiredForImport` and `OptionalForImport`, which, as named, only affect behavior when importing a resource by identity.
```

>  - `RequiredForImport` is set to true if an import block would need this information as part of the identity attribute to import the resource. Some attributes might be specified through other means e.g. the provider configuration and therefore won't be needed for import.

Author: ```suggestion
 - `RequiredForImport` is set to true if an import block would need this information from the practitioner to import the resource.
```

>   - `OptionalForImport` is set to true if an import block can provide this information as part of the identity attribute to import the resource. This can be data that would normally be the default of the configured provider running the import. This does not mean the data is omitted from the resource identity, import-only data is not supported.

Author: Rewording, also the supporting of storing `null` for an identity attribute means that import only is technically possible 😆 

```suggestion
  - `OptionalForImport` is set to true if an import block does not require this information from the practitioner to import the resource. This can be data that would normally be defaulted from the configured provider running the import.
```

> It is expected that exactly one of `RequiredForImport` or `OptionalForImport` is set to true.

Author: ```suggestion
It is expected that exactly one of `RequiredForImport` or `OptionalForImport` is set to true. Regardless of which option is used, the provider can decide exactly what data is stored in the identity during import, similar to `Computed` attributes in resource state.
```

> 3. Identity attributes are expected to be unique, immutable and fully known.
> 

Author: This is mentioned in one of the above suggestions (which specifically talks about the identity object, not the attributes)

```suggestion
```

Example of an identity schema for an aws_s3_bucket:
```go
var _ resource.ResourceWithIdentity = &S3BucketResource{} // new interface

func (r *S3BucketResource) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
	resp.TypeName = "aws_s3_bucket"
}

func (r *S3BucketResource) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
	resp.Schema = schema.Schema{
		// .. existing resource schema for aws_s3_bucket
	}
}

func (r *S3BucketResource) IdentitySchema(_ context.Context, _ resource.IdentitySchemaRequest, resp *resource.IdentitySchemaResponse) {
	resp.IdentitySchema = identityschema.Schema{
		Attributes: map[string]identityschema.Attribute{
			"bucket": identityschema.StringAttribute{
				RequiredForImport: true,
			},
			"account_id": identityschema.StringAttribute{
				OptionalForImport: true,
			},
			"partition": identityschema.StringAttribute{
				OptionalForImport: true,
			},
			"region": identityschema.StringAttribute{
				OptionalForImport: true,

			},
		},
	}
}

> ```

Author: This is a good start, before we can do anything, we must choose the attributes for an identity, step 1!

  "role_assignment_id": "/providers/Microsoft.Authorization/roleAssignments/00000000-0000-0000-0000-000000000000",
}
```

#### AzureAD

##### azuread_group
Example schema:
```go
{
  "object_id": {string, required_for_import:true},
}
```

Example identity data:
```go
{
  object_id = "00000000-0000-0000-0000-000000000000"
}
```

##### Azuread_group_member
Example schema:
```go
{
  "group_object_id": {string, required_for_import:true},
  "member_object_id": {string, required_for_import:true},
}
```

Example identity data:
```go
{
  group_object_id = "00000000-0000-0000-0000-000000000000"
  member_object_id = "11111111-1111-1111-1111-111111111111"
}
```
##### azuread_application
Example schema:
```go
{
  "object_id": {string, required_for_import:true},
}
```

Example identity data:
```go
{
  object_id = "00000000-0000-0000-0000-000000000000"
}
```
##### azuread_app_role_assignment
Example schema:
```go
{
  "principal_object_id": {string, required_for_import:true},
  "app_role_id": {string, required_for_import:true},
}
```

Example identity data:
```go
{
  principal_object_id = "00000000-0000-0000-0000-000000000000"
  app_role_id = "aaBBcDDeFG6h5JKLMN2PQrrssTTUUvWWxxxxxyyyzzz"
}

> ```

Author: I think we may have gotten wires crossed here, this is all information on how to retrieve an identity _schema_ from the CLI, whereas, the provider developer at this point is probably interested in "how/where do I [set identity data](https://github.com/hashicorp/terraform-provider-corner/blob/b06ba2acba9cc98cf58cda4a1ecd7239443c4c94/internal/framework5provider/identity_resource.go#L60-L63), how/where do I retrieve identity data (`Identity.Get`)"

Those would be in Go methods similar to the "Create" or "Read" sections of resource, but probably smaller:
- https://developer.hashicorp.com/terraform/plugin/framework/resources/create#define-create-method
- https://developer.hashicorp.com/terraform/plugin/framework/resources/read#define-read-method
